pex/pep_376.py

```diff
@@ -149,77 +418,144 @@ def read(
+        base_dir = os.path.relpath(site_packages, prefix_dir)
         return cls(
-            project_name=dist.project_name,
-            version=dist.version,
-            base_location=dist.location,
-            relative_path=relative_path,
-            data_dir=data_dir,
+            project_name=project_name,
+            version=version,
+            prefix_dir=prefix_dir,
+            rel_base_dir=base_dir,
+            relative_path=record_relative_path,
             metadata_listing=tuple(
-                path for path in listing if metadata_dir == os.path.dirname(path)
+                path for path in site_packages_listing if metadata_dir == os.path.dirname(path)
             ),
-            install_scheme=install_scheme,
         )
 
     project_name = attr.ib()  # type: str
     version = attr.ib()  # type: str
-    base_location = attr.ib()  # type: str
+    prefix_dir = attr.ib()  # type: str
+    rel_base_dir = attr.ib()  # type: str
     relative_path = attr.ib()  # type: str
-    _data_dir = attr.ib()  # type: str
     _metadata_listing = attr.ib()  # type: Tuple[str, ...]
-    install_scheme = attr.ib(default=InstallationScheme.TARGET)  # type: InstallationScheme.Value
 
     def _find_dist_info_file(self, filename):
         # type: (str) -> Optional[str]
-        return dist_metadata.find_dist_info_file(
+        metadata_file = dist_metadata.find_dist_info_file(
             project_name=self.project_name,
             version=self.version,
             filename=filename,
             listing=self._metadata_listing,
         )
+        if not metadata_file:
+            return None
+        return os.path.join(self.rel_base_dir, metadata_file)
 
-    def fixup_install(self):
-        # type: () -> None
-        """Fixes a wheel install to be reproducible."""
+    def fixup_install(
+        self,
+        exclude=(),  # type: Container[str]
+        interpreter=None,  # type: Optional[PythonInterpreter]
+    ):
+        # type: (...) -> InstalledWheel
+        """Fixes a wheel install to be reproducible and importable.
+
+        After fixed up, this RECORD can be used to re-install the wheel in a venv with `reinstall`.
 
-        modified_scripts = list(self._fixup_scripts())
-        self._fixup_record(modified_scripts=modified_scripts)
+        :param exclude: Any top-level items to exclude.
+        :param interpreter: The interpreter used to perform the wheel install.
+        """
+        self._fixup_scripts()
+        self._fixup_direct_url()
+
+        # The RECORD is unused in PEX zipapp mode and only needed in venv mode. Since it can contain
+        # relative path entries that differ between interpreters - notably pypy for Python < 3.8 has
+        # a custom scheme - we just delete the file and create it on-demand for venv re-installs.
+        os.unlink(os.path.join(self.prefix_dir, self.rel_base_dir, self.relative_path))
+
+        # An example of the installed wheel chroot we're aiming for:
```

Comment:
There is a _lot_ of path gymnastics in this PR. They all center around this final layout. The `pip install --prefix` scheme is the same as a venv layout:
```
venv/
  bin/
  include/
  lib/
    pythonX.Y/
      site-packages/
  <data>
```

The main trick is to use that scheme, since its non-buggy unlike `pip install --target`, but preserve the zipapp mode semantics by moving all the `site-packages` to the root and shifting everything remaining under `.prefix` since dotted names are never Python importable. At venv creation time this re-shuffle is reversed and a valid RECORD written out.

Reply:
Makes sense. I think this would be helpful to include in the comments, including the trick about `.prefix` being non-importable